crates/parcel/src/requests/asset_request.rs

```diff
@@ -0,0 +1,168 @@
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use parcel_core::cache::CacheRef;
+use parcel_core::plugin::AssetBuildEvent;
+use parcel_core::plugin::BuildProgressEvent;
+use parcel_core::plugin::InitialAsset;
+use parcel_core::plugin::ReporterEvent;
+use parcel_core::plugin::RunTransformContext;
+use parcel_core::plugin::TransformResult;
+use parcel_core::plugin::TransformationInput;
+use parcel_core::types::Asset;
+use parcel_core::types::AssetStats;
+use parcel_core::types::Dependency;
+use parcel_core::types::Environment;
+use parcel_core::types::FileType;
+use parcel_filesystem::FileSystemRef;
+
+use crate::plugins::Plugins;
+use crate::plugins::TransformerPipeline;
+use crate::request_tracker::{Request, RequestResult, RunRequestContext, RunRequestError};
+
+/// The AssetRequest runs transformer plugins on discovered Assets.
+/// - Decides which transformer pipeline to run from the input Asset type
+/// - Runs the pipeline in series, switching pipeline if the Asset type changes
+/// - Stores the final Asset source code in the cache, for access in packaging
+/// - Finally, returns the complete Asset and it's discovered Dependencies
+pub struct AssetRequest<'a> {
+  pub env: Arc<Environment>,
+  pub file_path: PathBuf,
+  pub code: Option<String>,
+  pub pipeline: Option<String>,
+  pub side_effects: bool,
+  // TODO: move the following to RunRequestContext
+  pub cache: CacheRef,
+  pub file_system: FileSystemRef,
+  pub plugins: Arc<Plugins<'a>>,
+}
+
+impl<'a> Hash for AssetRequest<'a> {
+  fn hash<H: Hasher>(&self, state: &mut H) {
+    // TODO: Just derive this once the contextual params are moved to RunRequestContext
+    self.file_path.hash(state);
+    self.code.hash(state);
+    self.pipeline.hash(state);
+    self.env.hash(state);
+    self.side_effects.hash(state);
+  }
```

Comment:
Missing plugins? Should invalidate the request when the config changes.

Reply:
Agreed, we will address this later either with invalidations or by including the transformers in the hash.